sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java

```diff
@@ -0,0 +1,980 @@
+          // In `ParquetToSparkSchemaConverter`, we map parquet UINT32 to our LongType.
+          // For unsigned int32, it stores as plain signed int32 in Parquet when dictionary
+          // fallbacks. We read them as long values.
+          return new UnsignedIntegerUpdater();
+        } else if (sparkType == DataTypes.ByteType) {
+          return new ByteUpdater();
+        } else if (sparkType == DataTypes.ShortType) {
+          return new ShortUpdater();
+        } else if (sparkType == DataTypes.DateType) {
+          if ("CORRECTED".equals(datetimeRebaseMode)) {
+            return new IntegerUpdater();
+          } else {
+            boolean failIfRebase = "EXCEPTION".equals(datetimeRebaseMode);
+            return new IntegerWithRebaseUpdater(failIfRebase);
+          }
+        }
+        throw constructConvertNotSupportedException(descriptor, sparkType);
+
+      case INT64:
+        // This is where we implement support for the valid type conversions.
+        if (sparkType == DataTypes.LongType || canReadAsLongDecimal(descriptor, sparkType)) {
+          return new LongUpdater(DecimalType.is32BitDecimalType(sparkType));
+        } else if (isUnsignedIntTypeMatched(64)) {
+          // In `ParquetToSparkSchemaConverter`, we map parquet UINT64 to our Decimal(20, 0).
+          // For unsigned int64, it stores as plain signed int64 in Parquet when dictionary
+          // fallbacks. We read them as decimal values.
+          return new UnsignedLongUpdater();
+        } else if (isTimestampTypeMatched(LogicalTypeAnnotation.TimeUnit.MICROS)) {
+          if ("CORRECTED".equals(datetimeRebaseMode)) {
+            return new LongUpdater(false);
+          } else {
+            boolean failIfRebase = "EXCEPTION".equals(datetimeRebaseMode);
+            return new LongWithRebaseUpdater(failIfRebase);
+          }
+        } else if (isTimestampTypeMatched(LogicalTypeAnnotation.TimeUnit.MILLIS)) {
+          if ("CORRECTED".equals(datetimeRebaseMode)) {
+            return new LongAsMicrosUpdater();
+          } else {
+            final boolean failIfRebase = "EXCEPTION".equals(datetimeRebaseMode);
+            return new LongAsMicrosRebaseUpdater(failIfRebase);
+          }
+        }
+        throw constructConvertNotSupportedException(descriptor, sparkType);
+
+      case FLOAT:
+        if (sparkType == DataTypes.FloatType) {
+          return new FloatUpdater();
+        }
+        throw constructConvertNotSupportedException(descriptor, sparkType);
+
+      case DOUBLE:
+        if (sparkType == DataTypes.DoubleType) {
+          return new DoubleUpdater();
+        }
+        throw constructConvertNotSupportedException(descriptor, sparkType);
+
+      case INT96:
+      case BINARY:
+        if (sparkType == DataTypes.StringType || sparkType == DataTypes.BinaryType ||
+          canReadAsBinaryDecimal(descriptor, sparkType)) {
+          return new BinaryUpdater();
+        } else if (sparkType == DataTypes.TimestampType) {
```

Comment:
Do we support reading parquet BINARY as timestamp before this PR?

Reply:
No we don't. For `BINARY` it will aways go to the first branch with `BinaryUpdater`. This follows the existing logic where we handle `INT96` and `BINARY` [in the same method](https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java#L686).

Reply:
What if the parquet field physical type is BINARY, and we read this parquet file with a user-specified type TimestampType?

Reply:
It will fail both with or without this PR, with the following error message:
```
org.apache.parquet.io.ParquetDecodingException: Failed to read 12 bytes
```

Similarly, for your other comment on `UINT32`, Spark will also happily read the int as unsigned long if user requested schema is `long` but Parquet type is `INT32`. 

I can fix both in this PR too.

Reply:
Yea let's fix them.